protocols/relay/src/v2/copy_future.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Helper to interconnect two substreams, connecting the receiver side of A with the sender side of
+//! B and vice versa.
+//!
+//! Inspired by [`futures::io::Copy`].
+
+use futures::future::Future;
+use futures::future::FutureExt;
+use futures::io::{AsyncBufRead, BufReader};
+use futures::io::{AsyncRead, AsyncWrite};
+use futures::ready;
+use futures_timer::Delay;
+use std::convert::TryInto;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+pub struct CopyFuture<S, D> {
```

Comment:
How hard would it be to add some unit tests for this? :grimacing:

Reply:
:D not hard. Added via 93691a81b48fc325f9a392e3dafbe2c2f6fd3226.

Testing:
    - Copying from a to b and vice versa.
    - Checking max bytes.
    - Checking max duration.


Please list any additional properties you would like to see tested.